### DISCUSSION TITLE
Cmake configure errors when using a cross compiled OpenDDS

### DISCUSSION THREAD

**Discussion by lc-gd** [OpenDDS/OpenDDS#4199] [Q&A] [answered]
host `Ubuntu 20.04`
using `OpenDDS-3.24.1`
cmake version `3.16.3`
GCC version (cross compiler) `10.2.0`

I'm having some difficulty getting a cmake project to successfully include openDDS that has been cross compiled for an arm target.

I based my process off of the raspberrypi build instructions with a few additions from the linux process, like `--prefix` for installing. I configured OpenDDS with the following command:
`./configure --prefix=/home/me/dds_export --target=linux-cross --target-compiler=/opt/Xilinx/Vitis/2021.1/gnu/aarch64/lin/aarch64-linux/bin/aarch64-linux-gnu-g++ --no-tests`
I then ran `make` and everything built successfully. From this point I was able to use the copying instructions to run the example applications on my target.

The problem I'm facing is when I now go to build the `Messenger` example with cmake.

Following the steps outlined in `docs/cmake.md` heading **Building the Developer's Guide Messenger Example**, the portion on using an installed OpenDDS.
I first run `make install` at the OpenDDS root. This fails perhaps expectedly with
`make: *** No rule to make target 'install'. Stop.`
I'm guessing this is because all the references to installing are assuming your target is also your host system so there is no `build` folder, no problem I instead cd into `build/target` and run `make install` again. this time it works and my `/home/me/dds_export` folder is populated with a `bin`, `include`, `lib`, and `share` directory.

Now from the target directory I cd into `DevGuideExamples/DCPS/Messenger`, create the `build` folder and cd into it. From here I run the command `cmake -DCMAKE_PREFIX_PATH="/home/me/dds_export" ..`. The cmake configuration fails at this point and gives me the following output:
```
-- The CXX compiler identification is GNU 10.2.0
-- Check for working CXX compiler: /opt/Xilinx/Vitis/2021.1/gnu/aarch64/lin/aarch64-linux/bin/aarch64-linux-gnu-g++
-- Check for working CXX compiler: /opt/Xilinx/Vitis/2021.1/gnu/aarch64/lin/aarch64-linux/bin/aarch64-linux-gnu-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is GNU 10.2.0
-- Check for working C compiler: /opt/Xilinx/Vitis/2021.1/gnu/aarch64/lin/aarch64-linux/bin/aarch64-linux-gnu-gcc
-- Check for working C compiler: /opt/Xilinx/Vitis/2021.1/gnu/aarch64/lin/aarch64-linux/bin/aarch64-linux-gnu-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
CMake Error at /home/me/dds_export/share/cmake/OpenDDS/init.cmake:73 (message):
  Missing required dependencies
  OPENDDS_DCPS_LIBRARY;OPENDDS_IDL;ACE_LIBRARY;ACE_GPERF;TAO_LIBRARY;TAO_IDL
Call Stack (most recent call first):
  /home/me/dds_export/share/cmake/OpenDDS/OpenDDSConfig.cmake:405 (_OPENDDS_RETURN_ERR)
  CMakeLists.txt:5 (find_package)


CMake Error at CMakeLists.txt:19 (OPENDDS_TARGET_SOURCES):
  Unknown CMake command "OPENDDS_TARGET_SOURCES".


-- Configuring incomplete, errors occurred!
See also "/home/me/Projects/OpenDDS-3.24.1/build/target/DevGuideExamples/DCPS/Messenger/build/CMakeFiles/CMakeOutput.log".
See also "/home/me/Projects/OpenDDS-3.24.1/build/target/DevGuideExamples/DCPS/Messenger/build/CMakeFiles/CMakeError.log".
```
It looks to me like the provided cmake config module cannot properly source those relative dependencies in the install folder. Tracing through it a bit specifically with the `TAO_IDL` dependency I think I notice a problem. This dependency name is looking for the program `tao_idl` using the `_tao_bin_hints` which is sourced from `TAO_BIN_DIR` which is set to `ACE_BIN_DIR` in the init.cmake which is set to `ACE_ROOT/bin` with `ACE_ROOT` resolving to the prefix path followed by `/share/ace` so the executable should be located for me at `/home/me/dds_export/share/ace/bin` but when I enter that directory I do not see the executable. 

Perhaps I'm misunderstanding something here, but it would appear that the install was incomplete and does not have all the tools needed for development. Additionally for the tao_idl executable I would think the version of that I want would need to be an x86 version as opposed to my arm target version, since that is where I am developing and the exe is used to generate .cpp/.h files that will then be used to build an executable for the target. Maybe I need to make install in `build/host` as well to get everything?

**Comment by lc-gd** [comment]
I've since run `make install` inside of `build/host` as well, this appears to have improved things somewhat with that list of missing dependencies shortening to just 3 libraries. I'm not sure how to resolve this however, and I don't know if that second install has overwritten some of the target install files or not
```
Missing required dependencies OPENDDS_DCPS_LIBRARY;ACE_LIBRARY;TAO_LIBRARY
```

**Comment by mitza-oci** [comment]
@iguessthislldo will probably have more recent/relevant advice on this - but my suggestion would be to skip `make install` and set up the CMake package to find what it needs (code generators, headers, libraries) based on the `build/*` tree.

**Reply by lc-gd** [reply]
I attempted to do that this way:

edit the Messenger/CmakeLists.txt with the following line before `find_package(...)`:
`list(APPEND CMAKE_PREFIX_PATH /home/me/Projects/OpenDDS-3.24.1/build/target)`
then I ran `source setenv.sh` in build/target (not sure if that's needed or not)
and then return to the Messenger build folder and attempt to configure cmake again `cmake ..`
which results in the same dependency error
```
Missing required dependencies OPENDDS_DCPS_LIBRARY;ACE_LIBRARY;TAO_LIBRARY
```

**Reply by mitza-oci** [reply]
Right, it's probably going to need some code changes to the CMake package.

**Reply by lc-gd** [reply]
I'm guessing they would just be things I can set in the `config.cmake` file? I'm not sure what I'd need to change however. I've been trying to trace those 3 library dependencies to see where their names get set, but outside of the reference to them in the `OpenDDSConfig.cmake` I can't locate any definition or macro that names them or appends `_LIBRARY` to something.

**Reply by lc-gd** [reply]
just an update. I added these 3 lines to the `OpenDDSConfig.cmake` just before the variables get dependency checked.
```
set(OPENDDS_DCPS_LIBRARY "/home/me/Projects/OpenDDS-3.24.1/build/target/dds/libOpenDDS_Dcps.so.3.24.1")
set(ACE_LIBRARY "/home/me/Projects/OpenDDS-3.24.1/build/target/ACE_wrappers/ace/libACE.so.6.5.19")
set(TAO_LIBRARY "/home/me/Projects/OpenDDS-3.24.1/build/target/ACE_wrappers/TAO/tao/libTAO.so.2.5.19")
```
running configure with these additions gets me farther than before, but I still run in to what looks like more of the same problems. It looks like I now can't reference components within the library?

```
-- Configuring done
CMake Error at CMakeLists.txt:37 (add_executable):
  Target "subscriber" links to target "OpenDDS::Dcps" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:37 (add_executable):
  Target "subscriber" links to target "OpenDDS::InfoRepoDiscovery" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at CMakeLists.txt:37 (add_executable):
  Target "subscriber" links to target "OpenDDS::Tcp" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:37 (add_executable):
  Target "subscriber" links to target "OpenDDS::Rtps" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:37 (add_executable):
  Target "subscriber" links to target "OpenDDS::Rtps_Udp" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?


CMake Error at CMakeLists.txt:37 (add_executable):
  Target "subscriber" links to target "OpenDDS::Dcps" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:31 (add_executable):
  Target "publisher" links to target "OpenDDS::Dcps" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:31 (add_executable):
  Target "publisher" links to target "OpenDDS::InfoRepoDiscovery" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at CMakeLists.txt:31 (add_executable):
  Target "publisher" links to target "OpenDDS::Tcp" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:31 (add_executable):
  Target "publisher" links to target "OpenDDS::Rtps" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:31 (add_executable):
  Target "publisher" links to target "OpenDDS::Rtps_Udp" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?


CMake Error at CMakeLists.txt:31 (add_executable):
  Target "publisher" links to target "OpenDDS::Dcps" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at CMakeLists.txt:19 (add_library):
  Target "messenger_idl" links to target "OpenDDS::Dcps" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

**Comment by iguessthislldo** [comment, marked as answer]
> I've been trying to trace those 3 library dependencies to see where their names get set, but outside of the reference to them in the OpenDDSConfig.cmake I can't locate any definition or macro that names them or appends _LIBRARY to something.

> running configure with these additions gets me farther than before, but I still run in to what looks like more of the same problems. It looks like I now can't reference components within the library?

The `_LIBRARY` variables shouldn't really be set manually, at least like that because they are part of bigger process that turns them into targets CMake can use. The gist is that libraries are passed to [`find_library`](https://github.com/OpenDDS/OpenDDS/blob/6d638395e340f483f5005264091c5a839a8050e9/cmake/OpenDDSConfig.cmake#L342) to create `*_LIBRARY_<CONFIG>` vars for each config (Debug and/or Release), then [`select_library_configuration`](https://github.com/OpenDDS/OpenDDS/blob/6d638395e340f483f5005264091c5a839a8050e9/cmake/OpenDDSConfig.cmake#L385C22-L385C22) uses that info to create the actual `*_LIBRARY` variable that the one check uses. After that the [`*_LIBRARY_<CONFIG>` variables are used to set properties](https://github.com/OpenDDS/OpenDDS/blob/6d638395e340f483f5005264091c5a839a8050e9/cmake/OpenDDSConfig.cmake#L424) on the the library target that users can use.

I got the 3.24 CMake package to cross compile, but it's basically a hack. I thought we might have supported cross compiling somewhat before, but it doesn't really look like it. This is for Android as I don't have a more appropriate Linux cross compiler that's set up at the moment.

It required a small change to `OpenDDSConfig.cmake` to tell it to not look for our libraries in the cross compiler's libraries. There's also a required change for Android to ignore rt that might also apply to your case if your target system doesn't use glibc, but I doubt that's the case.

```dif
diff --git a/cmake/OpenDDSConfig.cmake b/cmake/OpenDDSConfig.cmake
index b3f3e3af69..c41dfcd9d5 100644
--- a/cmake/OpenDDSConfig.cmake
+++ b/cmake/OpenDDSConfig.cmake
@@ -200,7 +200,7 @@ endmacro()
 
 if(UNIX)
   _OPENDDS_SYSTEM_LIBRARY(dl)
-  if(NOT APPLE)
+  if(NOT APPLE AND NOT ANDROID)
     _OPENDDS_SYSTEM_LIBRARY(rt)
   endif()
 elseif(MSVC)
@@ -339,7 +339,7 @@ function(opendds_find_our_libraries_for_config config suffix)
       string(TOUPPER ${lib} var_prefix)
       set(lib_var "${var_prefix}_LIBRARY_${config}")
 
-      find_library(${lib_var} "${lib_file_base}" HINTS "${lib_dir}")
+      find_library(${lib_var} "${lib_file_base}" HINTS "${lib_dir}" NO_CMAKE_FIND_ROOT_PATH)
       set(found_var "${var_prefix}_LIBRARY_FOUND")
       if(${lib_var})
         set(${found_var} TRUE PARENT_SCOPE)
```

Next I needed to override the [bin directories](https://github.com/OpenDDS/OpenDDS/blob/6d638395e340f483f5005264091c5a839a8050e9/cmake/init.cmake#L106-L157) so it pointed to the host tools. I was trying to do without editing files, so I went with this this configure command below. The important part is the `-D`s below `-DOpenDDS_DIR`. The `*_ROOT` variables allow overriding the other variables, specifically the [`*_BIN_DIR` ones](https://github.com/OpenDDS/OpenDDS/blob/6d638395e340f483f5005264091c5a839a8050e9/cmake/init.cmake#L116) that we want to point to the host tools. Ideally that's all we'd have to set, but the way it's written all these variables need to be set. They can be put in sets in `config.cmake` or your `CMakeLists.txt` before `find_package(OpenDDS)` if you want.

```
ACE_ROOT=$(realpath build/target/ACE_wrappers)
TAO_ROOT=$(realpath build/target/ACE_wrappers/TAO)
DDS_ROOT=$(realpath build/target)
cmake -S DevGuideExamples/DCPS/Messenger -B build-msg \
  -DANDROID_ABI=arm64-v8a -DANDROID_PLATFORM=android-29 \
  --toolchain $(realpath ../../cmake/android-ace-tao/android-ndk-r25c/build/cmake/android.toolchain.cmake) \
  -DOpenDDS_DIR=$(realpath build/target/cmake) \
  -DACE_ROOT=$ACE_ROOT \
  -DACE_INCLUDE_DIRS=$ACE_ROOT \
  -DACE_BIN_DIR=$(realpath build/host/ACE_wrappers/bin) \
  -DACE_LIB_DIR=$ACE_ROOT/lib \
  -DTAO_ROOT=$TAO_ROOT \
  -DTAO_INCLUDE_DIR=$TAO_ROOT \
  -DTAO_INCLUDE_DIRS=$TAO_ROOT\;$TAO_ROOT/orbsvcs \
  -DTAO_BIN_DIR=$(realpath build/host/ACE_wrappers/bin) \
  -DTAO_LIB_DIR=$ACE_ROOT/lib \
  -DDDS_ROOT=$DDS_ROOT \
  -DOPENDDS_INCLUDE_DIRS=$DDS_ROOT \
  -DOPENDDS_BIN_DIR=$(realpath build/host/bin) \
  -DOPENDDS_LIB_DIR=$DDS_ROOT/lib
```

These might or might not work for you since I think CMake toolchain files can set different options for how programs and libraries are supposed to be found.

This definitely needs to be improved. I have some support for cross-compilling in my current ongoing work to build OpenDDS itself with CMake, I think that can be extended to over MPC-built CMake using some of what I learned here.

**Reply by lc-gd** [reply]
Thanks for your assistance, this does in fact work and I was able to successfully cross compile the Messenger example project with cmake!

**Reply by iguessthislldo** [reply]
The proper support for all this has been merged (#4203) into master for the upcoming 3.26 release. Now all it will require is setting [`OPENDDS_HOST_TOOLS`](https://opendds.readthedocs.io/en/master/devguide/building/cmake.html#var-OPENDDS_HOST_TOOLS) and [`OPENDDS_ACE_TAO_HOST_TOOLS`](https://opendds.readthedocs.io/en/master/devguide/building/cmake.html#var-OPENDDS_ACE_TAO_HOST_TOOLS).
